### PR TITLE
add position action and lambda - tested

### PR DESCRIPTION
**Pull request by KristopherMackowiak**
## Description:

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/419

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by KristopherMackowiak** [comment]
+ doc
(https://github.com/esphome/esphome-docs/pull/419)

File: esphome/components/template/cover/__init__.py [hunk 1]
```diff
@@ -60,6 +62,15 @@ def to_code(config):
     cg.add(var.set_assumed_state(config[CONF_ASSUMED_STATE]))
     cg.add(var.set_restore_mode(config[CONF_RESTORE_MODE]))
     cg.add(var.set_has_position(config[CONF_HAS_POSITION]))
+    
+    if CONF_POSITION_ACTION in config:
+        yield automation.build_automation(var.get_position_trigger(), [(float, 'pos')],
+                                          config[CONF_POSITION_ACTION])
+        cg.add(var.set_has_position(True))
+    if CONF_POSITION in config:
```
Comment (glmnet) [inline esphome/components/template/cover/__init__.py]: shouldn't this be `CONF_POSITION_LAMBDA`

File: esphome/components/template/cover/template_cover.cpp [hunk 2]
```diff
@@ -86,9 +86,10 @@ void TemplateCover::control(const CoverCall &call) {
     } else if (pos == COVER_CLOSED) {
       this->close_trigger_->trigger();
       this->prev_command_trigger_ = this->close_trigger_;
+    } else {
+      this->position_trigger_->trigger(pos);
     }
 
-    this->position_trigger_->trigger(pos);
```
Comment (glmnet) [inline esphome/components/template/cover/template_cover.cpp]: I believe you'll still want an update on a fully open, fully close, is there a reason for not doing so?

**Review by glmnet — CHANGES_REQUESTED** [review]
Please add usage of this on the test1.yaml or another test yaml files there, you should use the added yaml options.

I believe you are still missing the template_cover.h file, the build should fail if you add the test.yaml

Also check the travis outputs, there are some code formatting issues needed to be solved (whitespace at the end of lines, extra line spaces, too long lines, etc.)

File: esphome/components/template/cover/__init__.py (hunk 1 above)
Comment (KristopherMackowiak) [reply esphome/components/template/cover/__init__.py]: you're right. I made a mistake copying my changes (I don't have a git installed and I did it on the website)

File: esphome/components/template/cover/template_cover.cpp (hunk 2 above)
Comment (KristopherMackowiak) [reply esphome/components/template/cover/template_cover.cpp]: it's good here. without 'else', the position change code would execute during the open and close actions. we want to have different actions for all cases

**Comment by KristopherMackowiak** [comment]
@glmnet in the template_cover.h I don't have missing objects. E.g. position trigger existed but was nowhere used. Is everything right now? can you make a merge?

**Comment by glmnet** [comment]
I believe there is a conceptual error here.

`position_lambda:`

Position is the main "state" of the cover itself, different from e.g. `tilt`

There is plain `lambda`, you should use this lambda to return the current position, i.e. this is supposed to be used, and makes `position_lambda` redundant.

Actually you are using `set_state_lambda` and that's why you did not modify the .h file.

**Comment by glmnet** [comment]
Also try to keep the things ordered, e.g. put the 

`if CONF_POSITION_ACTION in config:`

after the stop action, before the tilt action, etc.

**Comment by KristopherMackowiak** [comment]
you are right that position lambda is unnecessary. I removed it. I miss positions action very much.

about order:
`if CONF_POSITION_ACTION in config:`
in the last release the (hash needless) parameter 'has_position' has been added. that's why I have to execute my code after setting this parameter.

**Comment by glmnet** [comment]
I see, it's just that I don't like a set_ being called twice (once with the default, then with your constan true)

What about something like this:

````
    if CONF_TILT_ACTION in config:
        yield automation.build_automation(var.get_tilt_trigger(), [(float, 'tilt')],
                                          config[CONF_TILT_ACTION])
        cg.add(var.set_has_tilt(True))
    if CONF_POSITION_ACTION in config:
        yield automation.build_automation(var.get_position_trigger(), [(float, 'pos')],
                                          config[CONF_POSITION_ACTION])
        cg.add(var.set_has_position(True))
    else:
        cg.add(var.set_has_position(config[CONF_HAS_POSITION]))

    if CONF_TILT_LAMBDA in config:
        tilt_template_ = yield cg.process_lambda(config[CONF_TILT_LAMBDA], [],
                                                 return_type=cg.optional.template(float))
        cg.add(var.set_tilt_lambda(tilt_template_))
    cg.add(var.set_optimistic(config[CONF_OPTIMISTIC]))
    cg.add(var.set_assumed_state(config[CONF_ASSUMED_STATE]))
    cg.add(var.set_restore_mode(config[CONF_RESTORE_MODE]))
````

Maybe you should also state in the docs that if `position_action` is set it will cause `has_position` to true.

**Comment by KristopherMackowiak** [comment]
I hope that now all is ok, becouse i flashed esphome in my blinds and wait for this change.

**Review by glmnet — APPROVED** [review]

**Pull request merged**